src/main/java/com/tabnine/Initializer.kt

```diff
@@ -0,0 +1,88 @@
+package com.tabnine
+
+import com.intellij.ide.plugins.PluginInstaller
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.application.PreloadingActivity
+import com.intellij.openapi.components.ServiceManager
+import com.intellij.openapi.diagnostic.Logger
+import com.intellij.openapi.progress.ProgressIndicator
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.startup.StartupActivity
+import com.tabnine.capabilities.CapabilitiesService
+import com.tabnine.config.Config
+import com.tabnine.dialogs.Dialogs.showRestartDialog
+import com.tabnine.dialogs.TabnineEnterpriseUrlDialogWrapper
+import com.tabnine.general.DependencyContainer
+import com.tabnine.general.StaticConfig
+import com.tabnine.general.Utils
+import com.tabnine.lifecycle.BinaryNotificationsLifecycle
+import com.tabnine.lifecycle.BinaryPromotionStatusBarLifecycle
+import com.tabnine.lifecycle.BinaryStateService
+import com.tabnine.lifecycle.TabnineUpdater
+import com.tabnine.logging.initTabnineLogger
+import com.tabnine.notifications.ConnectionLostNotificationHandler
+import com.tabnine.userSettings.AppSettingsState.Companion.instance
+import java.util.concurrent.atomic.AtomicBoolean
+
+class Initializer : PreloadingActivity(), StartupActivity {
+    private var binaryNotificationsLifecycle: BinaryNotificationsLifecycle? = null
+    private var binaryPromotionStatusBarLifecycle: BinaryPromotionStatusBarLifecycle? = null
+    override fun preload(indicator: ProgressIndicator) {
+        initialize()
+    }
+
+    override fun runActivity(project: Project) {
+        initialize()
+    }
+
+    private fun initialize() {
+        val shouldInitialize = !(initialized.getAndSet(true) || ApplicationManager.getApplication().isUnitTestMode)
+        if (shouldInitialize) {
+            Logger.getInstance(javaClass)
+                .info(
+                    "Initializing for ${Config.CHANNEL}, plugin id = ${StaticConfig.TABNINE_PLUGIN_ID_RAW}"
+                )
+            connectionLostNotificationHandler.startConnectionLostListener()
+            ServiceManager.getService(BinaryStateService::class.java).startUpdateLoop()
+            initTabnineLogger()
+            if (!Config.IS_ON_PREM) {
+                binaryNotificationsLifecycle = DependencyContainer.instanceOfBinaryNotifications()
+                binaryPromotionStatusBarLifecycle = DependencyContainer.instanceOfBinaryPromotionStatusBar()
```

Comment:
i think its unnecessary to initialize these fields here and not in the constructor (just like `connectionLostNotificationHandler`) - I think it's a small mistake that stayed with us for a long time. When you convert it to Kotlin, this mistake becomes clearer.

I'd change these fields to be initialized in the beginning and ensure everything is working by building a local plugin for production and ensuring you receive notifications and such.